src/NumericalAlgorithms/LinearOperators/DefiniteIntegral.cpp

```diff
@@ -49,9 +35,75 @@ double definite_integral<1>(const DataVector& integrand,
-                                     const Mesh<3>&) noexcept;
+template <>
+double definite_integral<2>(const DataVector& integrand,
+                            const Mesh<2>& mesh) noexcept {
+  ASSERT(integrand.size() == mesh.number_of_grid_points(),
+         "num_grid_points = " << mesh.number_of_grid_points()
+                              << ", integrand size = " << integrand.size());
+  const auto sliced_meshes = mesh.slices();
+  const size_t x_size = sliced_meshes[0].number_of_grid_points();
+  const size_t x_final = x_size - x_size % 2;
+  const size_t y_size = sliced_meshes[1].number_of_grid_points();
+  const double* const w_x =
+      Spectral::quadrature_weights(sliced_meshes[0]).data();
+  const double* const w_y =
+      Spectral::quadrature_weights(sliced_meshes[1]).data();
+
+  double result = 0.0;
+  for (size_t j = 0; j < y_size; ++j) {
+    const size_t offset = j * x_size;
+    for (size_t i = 0; i < x_final; i += 2) {
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+      result += w_y[j] * w_x[i] * integrand[i + offset] +
+                w_y[j] * w_x[i + 1] * integrand[i + 1 + offset];  // NOLINT
+    }
+    for (size_t i = x_final; i < x_size; ++i) {
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+      result += w_y[j] * w_x[i] * integrand[i + offset];
+    }
+  }
+  return result;
+}
+
+template <>
+double definite_integral<3>(const DataVector& integrand,
+                            const Mesh<3>& mesh) noexcept {
+  ASSERT(integrand.size() == mesh.number_of_grid_points(),
+         "num_grid_points = " << mesh.number_of_grid_points()
+                              << ", integrand size = " << integrand.size());
+  const auto sliced_meshes = mesh.slices();
+  const size_t x_size = sliced_meshes[0].number_of_grid_points();
+  const size_t x_final = x_size - x_size % 2;
+  const size_t y_size = sliced_meshes[1].number_of_grid_points();
+  const size_t z_size = sliced_meshes[2].number_of_grid_points();
+  const double* const w_x =
+      Spectral::quadrature_weights(sliced_meshes[0]).data();
+  const double* const w_y =
+      Spectral::quadrature_weights(sliced_meshes[1]).data();
+  const double* const w_z =
+      Spectral::quadrature_weights(sliced_meshes[2]).data();
+
+  double result = 0.0;
+  for (size_t k = 0; k < z_size; ++k) {
+    for (size_t j = 0; j < y_size; ++j) {
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+      const double prod = w_z[k] * w_y[j];
+      const size_t offset = x_size * (j + y_size * k);
+      // Unrolling at 2 gives better performance on AVX machines (the most
+      // common in 2018). The stride will probably need to be updated as
+      // hardware changes. Using a single loop is ~15% faster if x_size < 4.
+      for (size_t i = 0; i < x_final; i += 2) {
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        result += prod * w_x[i] * integrand[i + offset] +
```

Comment:
does factoring out the prod to `prod*(w_x[i] * integrand[i + offset] + w_x[i + 1] * integrand[i + 1 + offset])` bump the needle at all?

Reply:
To put some perspective on this, we're talking about potential differences of 33ns to 32ns per evaluation. This is already at the point of unless literally all we're doing is definite integrals it's not going to make a difference. It's also at the point where I can't really tell what's faster anymore because it's in the noise because I don't have a system that is running a super minimal OS with nothing else going on.

Reply:
All good - I do understand a point of diminishing returns, I just wanted to check anything easy to tweak while you've got a nicely working benchmark function. If you're confident you've reached the performance asymptote on these, I'm happy.

Reply:
:+1: yes, I think anything left will be hardware dependent (e.g. running with AVX512 instead of AVX2, etc.) which we will need to tweak for different computers if we notice performance issues :)